stackslib/src/chainstate/stacks/miner.rs

```diff
@@ -2661,3 +2661,244 @@ impl StacksBlockBuilder {
         Ok((block, consumed, size))
     }
 }
+
+/// Represents a block proposed to the `v2/block_proposal` endpoint for validation
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockProposal {
+    pub parent_consensus_hash: ConsensusHash,
+    /// Block
+    pub block: StacksBlock,
+    /// These are all the microblocks that the proposed block
+    /// will confirm.
+    pub microblocks_confirmed: Vec<StacksMicroblock>,
+    /// This refers to the burn block that was the current tip
+    ///  at the time this proposal was constructed. In most cases,
+    ///  if this proposal is accepted, it will be "mined" in the next
+    ///  burn block.
+    pub burn_tip: BurnchainHeaderHash,
+    /// This refers to the burn block that was the current tip
+    ///  at the time this proposal was constructed. In most cases,
+    ///  if this proposal is accepted, it will be "mined" in the next
+    ///  burn block.
+    pub burn_tip_height: u32,
+    /// Mainnet flag
+    pub is_mainnet: bool,
+    /// This is the public key hash which will be used to sign subsequent
+    ///  microblocks.
+    pub microblock_pubkey_hash: Hash160,
+    /// This is the total burn amount up to this block, used in the
+    ///  Stacks header
+    pub total_burn: u64,
+}
+
+/// This enum is used to supply a `reason_code` for validation
+///  rejection responses. This is serialized as an enum with string
+///  type (in jsonschema terminology).
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub enum ValidateRejectCode {
+    BadBlockHash,
+    BadTransaction,
+    InvalidBlock,
+    ChainstateError,
+    UnknownParent,
+}
+
+/// This enum is used for serializing the response to block
+/// proposal validation.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+#[serde(tag = "Result")]
+pub enum BlockValidateResponse {
+    Ok(BlockValidateOk),
+    Reject(BlockValidateReject),
+}
+
+/// A response for block proposal validation
+///  that the stacks-node thinks is acceptable.
+#[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
+pub struct BlockValidateOk {
+    pub block: StacksBlock,
```

Comment:
Just flagging here that this will eventually need to be ported to `NakamotoBlock`.

Reply:
Is there any reason this needs to work for non-Nakamoto blocks?

Reply:
No -- it doesn't need to work for non-Nakamoto blocks. Updating this to be `NakamotoBlock` would require re-basing this off of `next`, though.